File: roboco/api/schemas/video.py
"""Schemas for the video engine's on-demand request + CEO approval surface."""

from datetime import datetime
from typing import TYPE_CHECKING, Any
from uuid import UUID

from pydantic import BaseModel, Field

from roboco.api.schemas.project_fields import task_project_fields
from roboco.foundation.policy.content import markers
from roboco.foundation.policy.content.markers import MAX_VIDEO_RENDER_ATTEMPTS
from roboco.services.video_post_service import MAX_TIKTOK_CAPTION_CHARS
from roboco.services.x_client import MAX_TWEET_CHARS

if TYPE_CHECKING:
    from roboco.db.tables import TaskTable


class VideoRequestBody(BaseModel):
    """The CEO's on-demand video brief, scoped to a specific project."""

    occasion: str = Field(..., min_length=1)
    brief: str = Field(..., min_length=1)
    platforms: list[str] = Field(..., min_length=1)
    project_id: UUID


class VideoRequestResponse(BaseModel):
    """The outcome of an on-demand video request."""

    status: str  # "opened" | "disabled" | "not_opened"
    task_id: str | None = None
    detail: str


class VideoPostResponse(BaseModel):
    """One held video_post draft (rendered clip) awaiting the CEO."""

    task_id: str
    source: str  # "video_post"
    title: str
    status: str
    occasion: str
    script: str
    platforms: list[str]
    x_caption: str | None = None
    tiktok_caption: str | None = None
    reject_reason: str | None = None
    mp4_paths: dict[str, str] = Field(default_factory=dict)
    source_task_id: str | None = None  # the authoring task this draft rendered from
    project_slug: str | None = None
    project_name: str | None = None


class VideoPostApproveRequest(BaseModel):
    """Approve a draft, optionally overwriting one or both captions first."""

    x_caption: str | None = Field(default=None, max_length=MAX_TWEET_CHARS)
    tiktok_caption: str | None = Field(
        default=None, max_length=MAX_TIKTOK_CAPTION_CHARS
    )


class VideoPostExecuteResponse(BaseModel):
    """The outcome of an approve call."""

    status: str
    posted: dict[str, str]
    detail: str


class VideoPostRejectRequest(BaseModel):
    """The CEO's reason for declining a draft."""

    reason: str = Field(min_length=4)


class VideoPostHistoryResponse(BaseModel):
    """One acted-on video_post draft (posted or rejected) — the CEO's
    history view."""

    task_id: str
    source: str  # "video_post"
    title: str
    status: str  # "completed" | "cancelled"
    occasion: str
    script: str
    platforms: list[str]
    x_caption: str | None = None
    tiktok_caption: str | None = None
    reject_reason: str | None = None
    posted: dict[str, str] = Field(default_factory=dict)  # platform -> posted id
    acted_at: datetime
    source_task_id: str | None = None  # the authoring task this draft rendered from
    project_slug: str | None = None
    project_name: str | None = None


class VideoPipelineItemResponse(BaseModel):
    """One in-flight source=video item — the Social page's pipeline-strip
    basis. Spans the authoring task's whole pre-post lifecycle: any
    non-terminal delivery status, then the render loop's retry/failure
    states. composition_id/render_status/render_attempts/render_error live
    in the orchestration_markers.video_draft JSON, not a column."""

    task_id: str
    title: str
    occasion: str
    status: str
    pr_number: int | None = None
    composition_id: str | None = None
    render_status: str | None = None
    render_attempts: int = 0
    max_attempts: int = MAX_VIDEO_RENDER_ATTEMPTS
    render_error: str | None = None
    project_slug: str | None = None
    project_name: str | None = None


class PreviewFrameResponse(BaseModel):
    """One extracted ``request_render`` preview frame — index/timestamp decoded
    server-side from the sidecar's self-describing filename."""

    index: int
    file: str
    timestamp_seconds: float


class VideoPreviewFramesResponse(BaseModel):
    """A video-authoring task's ``request_render`` preview — the CEO's only
    look at the rendered artifact before the post-completion render loop
    produces the real MP4 (an ``awaiting_ceo_approval`` task otherwise has
    nothing to watch). Frames keyed by orientation; an absent/empty key was
    never rendered. Metadata (composition_id, duration_seconds, head_sha,
    dirty, rendered_at) comes from the last ``render_preview`` marker."""

    task_id: str
    composition_id: str | None = None
    duration_seconds: float | None = None
    head_sha: str | None = None
    dirty: bool | None = None
    rendered_at: str | None = None
    frames: dict[str, list[PreviewFrameResponse]] = Field(default_factory=dict)


class TikTokCredentialsStatus(BaseModel):
    """Whether the four OAuth2 secrets are stored. Never the secrets themselves."""

    has_credentials: bool


class TikTokCredentialsSetRequest(BaseModel):
    """Set (or, if all four are empty, clear) the four OAuth2 secrets."""

    client_key: str = Field(default="")
    client_secret: str = Field(default="")
    access_token: str = Field(default="")
    refresh_token: str = Field(default="")


def status_value(task: "TaskTable") -> str:
    raw = task.status
    return raw.value if hasattr(raw, "value") else str(raw)


def task_to_video_post_response(task: "TaskTable") -> VideoPostResponse:
    draft = markers.get_video_draft(task) or {}
    project_slug, project_name = task_project_fields(task)
    return VideoPostResponse(
        task_id=str(task.id),
        source=task.source,
        title=task.title,
        status=status_value(task),
        occasion=str(draft.get("occasion") or ""),
        script=str(draft.get("script") or ""),
        platforms=list(draft.get("platforms") or []),
        x_caption=draft.get("x_caption"),
        tiktok_caption=draft.get("tiktok_caption"),
        reject_reason=markers.get_video_reject_reason(task),
        mp4_paths=dict(draft.get("mp4_paths") or {}),
        source_task_id=draft.get("source_task_id"),
        project_slug=project_slug,
        project_name=project_name,
    )


def task_to_pipeline_item(task: "TaskTable") -> VideoPipelineItemResponse:
    draft = markers.get_video_draft(task) or {}
    project_slug, project_name = task_project_fields(task)
    return VideoPipelineItemResponse(
        task_id=str(task.id),
        title=task.title,
        occasion=str(draft.get("occasion") or ""),
        status=status_value(task),
        pr_number=task.pr_number,
        composition_id=draft.get("composition_id"),
        render_status=draft.get("render_status"),
        render_attempts=int(draft.get("render_attempts", 0)),
        render_error=draft.get("render_error"),
        project_slug=project_slug,
        project_name=project_name,
    )


def posted_ids(draft: dict[str, Any]) -> dict[str, str]:
    """Every ``{platform}_posted_id`` key stamped by approve, keyed by
    platform (e.g. ``{"x": "..", "tiktok": ".."}``)."""
    suffix = "_posted_id"
    return {
        k[: -len(suffix)]: str(v) for k, v in draft.items() if k.endswith(suffix) and v
    }


def task_to_video_post_history_response(task: "TaskTable") -> VideoPostHistoryResponse:
    draft = markers.get_video_draft(task) or {}
    project_slug, project_name = task_project_fields(task)
    return VideoPostHistoryResponse(
        task_id=str(task.id),
        source=task.source,
        title=task.title,
        status=status_value(task),
        occasion=str(draft.get("occasion") or ""),
        script=str(draft.get("script") or ""),
        platforms=list(draft.get("platforms") or []),
        x_caption=draft.get("x_caption"),
        tiktok_caption=draft.get("tiktok_caption"),
        reject_reason=markers.get_video_reject_reason(task),
        posted=posted_ids(draft),
        acted_at=task.updated_at or task.created_at,
        source_task_id=draft.get("source_task_id"),
        project_slug=project_slug,
        project_name=project_name,
    )

File: panel/src/lib/api/video.ts
import api, { API_URL } from "./client";

// ---------------------------------------------------------------------------
// Video engine — held rendered-clip drafts (script + 9:16/1:1 MP4 + per-
// platform captions) the CEO previews, edits, and approves (posts to X /
// TikTok) or rejects in the panel. Nothing posts until an explicit approve;
// TikTok credentials are write-only (the API never returns the stored
// secrets). Mirrors lib/api/x.ts.
// ---------------------------------------------------------------------------

export type VideoCut = "vertical" | "square";

export interface VideoPost {
  task_id: string;
  source: string; // "video_post"
  title: string;
  status: string;
  occasion: string;
  script: string;
  platforms: string[]; // subset of "x" | "tiktok"
  x_caption?: string | null;
  tiktok_caption?: string | null;
  reject_reason?: string | null;
  mp4_paths?: Record<string, string>;
  source_task_id?: string | null; // the authoring task this draft rendered from
  // The following mirror the shared `video_draft` marker's render idempotency
  // keys (same fields as VideoPipelineItem) — optional because a draft may
  // predate the backend exposing them on this response. Undefined reads as
  // "not stale" everywhere below, so the re-render control degrades safely.
  composition_id?: string | null;
  render_status?: string | null; // null | "rendered" | "failed"
  project_slug?: string | null;
  project_name?: string | null;
}

// One in-flight source=video authoring task — GET /video/pipeline
// (roboco/api/routes/video.py). Spans claim through the render loop's
// retry/failure states; a rendered task drops out (it's visible via
// VideoPost/listPosts instead).
export interface VideoPipelineItem {
  task_id: string;
  title: string;
  occasion: string;
  status: string;
  pr_number: number | null;
  composition_id: string | null;
  render_status: string | null; // null (pending/retrying) | "rendered" | "failed"
  render_attempts: number;
  max_attempts: number;
  render_error: string | null;
  project_slug?: string | null;
  project_name?: string | null;
}

export interface VideoPostExecuteResult {
  status: string; // posted | posted_partial | post_failed | already_posted | ...
  posted: Record<string, string>; // platform -> posted id
  detail: string;
}

// One acted-on draft (posted or rejected) — the CEO's history view.
export interface VideoPostHistoryEntry {
  task_id: string;
  source: string; // "video_post"
  title: string;
  status: string; // "completed" | "cancelled"
  occasion: string;
  script: string;
  platforms: string[];
  x_caption?: string | null;
  tiktok_caption?: string | null;
  reject_reason?: string | null;
  posted: Record<string, string>; // platform -> posted id
  acted_at: string;
  source_task_id?: string | null; // the authoring task this draft rendered from
  project_slug?: string | null;
  project_name?: string | null;
}

export interface VideoRequestResult {
  status: string; // "opened" | "disabled" | "not_opened"
  task_id?: string | null;
  detail: string;
}

export interface TikTokCredentialsStatus {
  has_credentials: boolean;
}

// One extracted request_render preview frame — GET /video/preview-frames/
// {task_id} (roboco/api/routes/video.py). index/timestamp decoded server-side
// from the sidecar's self-describing filename.
export interface PreviewFrame {
  index: number;
  file: string;
  timestamp_seconds: number;
}

// A video-authoring task's request_render preview — the CEO's only look at
// the rendered artifact before the post-completion render loop produces the
// real MP4 (an awaiting_ceo_approval task otherwise has nothing to watch).
// frames keyed by orientation; an absent/empty key was never rendered.
export interface VideoPreviewFrames {
  task_id: string;
  composition_id: string | null;
  duration_seconds: number | null;
  head_sha: string | null;
  dirty: boolean | null;
  rendered_at: string | null;
  frames: Partial<Record<VideoCut, PreviewFrame[]>>;
}

// GET /video/posts/{id}/media (roboco/api/routes/video.py) serves one
// rendered MP4 cut; VideoPost.mp4_paths (above) carries the server-side
// paths per cut. This builds that route's URL — but a native
// <video src> pointed straight at it 401s (a plain <video> GET carries none
// of axios's X-Agent-ID/X-Agent-Role headers), so the panel instead fetches
// the bytes via videoApi.getMediaBlob (axios) and drives <video> off an
// object URL. Kept for any direct-link use (e.g. opening the raw file).
export function videoMediaUrl(taskId: string, cut: VideoCut): string {
  return `${API_URL}/video/posts/${taskId}/media?cut=${cut}`;
}

// GET /video/preview/{task_id}/{file_path:path} (roboco/api/routes/video.py)
// serves a video-authoring task's composition HTML + sibling assets directly
// from the project's merged read-clone, with iframe-permitting headers — the
// panel's live composition preview embeds this URL as an <iframe src>
// directly (unlike the MP4 media route, no auth-header workaround needed).
export function compositionPreviewUrl(
  authoringTaskId: string,
  compositionId: string,
  cut: VideoCut,
): string {
  const filePath = `motion/compositions/${compositionId}/${cut}.html`;
  return `${API_URL}/video/preview/${authoringTaskId}/${filePath}`;
}

// GET /video/preview-frames/{task_id}/{orientation}/{filename} streams one
// frame's PNG bytes. Same auth-header problem as the MP4 media route (a
// plain <img src> GET carries none of axios's headers), so the panel fetches
// via videoApi.getPreviewFrameBlob instead — kept for direct-link use.
export function previewFrameUrl(
  taskId: string,
  cut: VideoCut,
  file: string,
): string {
  return `${API_URL}/video/preview-frames/${taskId}/${cut}/${encodeURIComponent(file)}`;
}

export const videoApi = {
  listPosts: async (): Promise<VideoPost[]> => {
    const { data } = await api.get<VideoPost[]>("/video/posts");
    return data;
  },
  // Every in-flight source=video authoring task — the pipeline strip's basis.
  listPipeline: async (): Promise<VideoPipelineItem[]> => {
    const { data } = await api.get<VideoPipelineItem[]>("/video/pipeline");
    return data;
  },
  // Posted or rejected drafts, newest-acted-first. Fixed default limit (50);
  // no "load more" — pass a higher limit if the panel ever needs it.
  listHistory: async (limit = 50): Promise<VideoPostHistoryEntry[]> => {
    const { data } = await api.get<VideoPostHistoryEntry[]>(
      "/video/posts/history",
      { params: { limit } },
    );
    return data;
  },
  // Fetches the rendered cut as a Blob via axios (carrying the auth headers
  // a plain <video src> GET can't) so the caller can drive <video> off an
  // object URL instead of pointing it at the route directly.
  getMediaBlob: async (taskId: string, cut: VideoCut): Promise<Blob> => {
    const { data } = await api.get<Blob>(`/video/posts/${taskId}/media`, {
      params: { cut },
      responseType: "blob",
    });
    return data;
  },
  approve: async (
    taskId: string,
    captions?: { x_caption?: string; tiktok_caption?: string },
  ): Promise<VideoPostExecuteResult> => {
    const { data } = await api.post<VideoPostExecuteResult>(
      `/video/posts/${taskId}/approve`,
      captions ?? {},
    );
    return data;
  },
  reject: async (taskId: string, reason: string): Promise<VideoPost> => {
    const { data } = await api.post<VideoPost>(
      `/video/posts/${taskId}/reject`,
      { reason },
    );
    return data;
  },
  requestVideo: async (body: {
    occasion: string;
    brief: string;
    platforms: string[];
    project_id: string;
  }): Promise<VideoRequestResult> => {
    const { data } = await api.post<VideoRequestResult>("/video/request", body);
    return data;
  },
  // POST /video/pipeline/{task_id}/rerender (roboco/api/routes/video.py) —
  // clears the authoring task's render idempotency keys so the render loop
  // re-picks it up. taskId is the *authoring* task (VideoPost.source_task_id),
  // not the held draft's own task_id.
  rerender: async (authoringTaskId: string): Promise<void> => {
    await api.post(`/video/pipeline/${authoringTaskId}/rerender`);
  },
  // The task's request_render preview frames, per orientation. Callers
  // treat a 404 (nothing rendered yet) as "no preview" rather than an error.
  getPreviewFrames: async (taskId: string): Promise<VideoPreviewFrames> => {
    const { data } = await api.get<VideoPreviewFrames>(
      `/video/preview-frames/${taskId}`,
    );
    return data;
  },
  // Fetches one frame's PNG bytes as a Blob (carrying the auth headers a
  // plain <img src> GET can't) — mirrors getMediaBlob's object-URL pattern.
  getPreviewFrameBlob: async (
    taskId: string,
    cut: VideoCut,
    file: string,
  ): Promise<Blob> => {
    const { data } = await api.get<Blob>(
      `/video/preview-frames/${taskId}/${cut}/${encodeURIComponent(file)}`,
      { responseType: "blob" },
    );
    return data;
  },
  getCredentialsStatus: async (): Promise<TikTokCredentialsStatus> => {
    const { data } = await api.get<TikTokCredentialsStatus>(
      "/tiktok/credentials",
    );
    return data;
  },
  setCredentials: async (creds: {
    client_key: string;
    client_secret: string;
    access_token: string;
    refresh_token: string;
  }): Promise<TikTokCredentialsStatus> => {
    const { data } = await api.post<TikTokCredentialsStatus>(
      "/tiktok/credentials",
      creds,
    );
    return data;
  },
};
